api.yaml

```diff
@@ -15,10 +15,111 @@ info:
+  /componentsusers:
     get:
+      summary: List all components and possible users
+      operationId: listComponentUsers
       responses:
         '200':
-          description: OK
+          description: >-
+            List of all the potential users and components with existing
+            subscribers and owners.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/ComponentsUsers'
+  /components/{componentId}/users/{userId}:
+    put:
+      summary: Add a user to a component
+      operationId: addUserToComponent
+      parameters:
+        - in: path
+          name: componentId
+          schema:
+            type: integer
+          required: true
+          description: Component ID
+        - in: path
+          name: userId
+          schema:
+            type: integer
+          required: true
+          description: User ID
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ComponentUsersRequest'
+      responses:
+        '200':
+          description: "Success"
+    delete:
+      summary: Remove a user from a component
+      operationId: removeUserFromComponent
+      parameters:
+        - in: path
+          name: componentId
+          schema:
+            type: integer
+          required: true
+          description: Component ID
+        - in: path
+          name: userId
+          schema:
+            type: integer
+          required: true
+          description: User ID
+      responses:
+        '200':
+          description: "Success"
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/ComponentUsersRequest'
+components:
+  schemas:        
+    ComponentsUsers:
+      properties:
+        users:
+          type: array
+          items:
+            type: object
+            properties:
+              id:
+                type: integer
+              name:
+                type: string
+              email:
+                type: string
+            required:
+              - id
+              - name
+              - email
+        components:
+          type: array
+          items:
+            type: object
+            properties:
+              id:
+                type: string
+              name:
+                type: string
+              subscriber_ids:
+                type: array
+                items:
+                  type: integer
+              owner_ids:
+                type: array
+                items:
+                  type: integer
+            required:
+              - id
+              - name
+
+    ComponentUsersRequest:
+      description: "Traits about the user in relation to the component"
+      properties:
+        owner:
+          type: boolean
+          description: Impacts this user's ownership. For PUT, add ownership. For DELETE, remove ownership.
```

Comment:
I'm concerned that this yaml file will be really big and hard to maintain.  The YAML syntax uses a ton of vertical space and there is no include directive.  Also, I wonder if indentation errors could be hard to track down.

I don't like the idea of layering on another code generation step, but what do you think about creating a simple python script (that would allow us to use python dicts or dataclasses, and python imports) that generates the api.yaml file (or a pretty-printed JSON file)?

Reply:
I figured out how to split the YAML into multiple files. More details here: https://github.com/GoogleChrome/chromium-dashboard/pull/2769#discussion_r1126956635

This should prevent the openapi file from growing too large. Also, with that we don't need to write our own code generation to build the final YAML.